python/ray/ml/train/trainer.py

```diff
@@ -0,0 +1,71 @@
+import abc
+import logging
+from typing import Dict, Union, Callable, Optional, TYPE_CHECKING
+
+from ray.ml.preprocessor import Preprocessor
+from ray.ml.checkpoint import Checkpoint
+from ray.ml.result import Result
+from ray.ml.config import ScalingConfig, RunConfig
+from ray.tune.trainable import ConvertibleToTrainable
+from ray.util import PublicAPI
+
+if TYPE_CHECKING:
+    from ray.data import Dataset
+
+# A type representing either a ray.data.Dataset or a function that returns a
+# ray.data.Dataset and accepts no arguments.
+GenDataset = Union["Dataset", Callable[[], "Dataset"]]
+
+
+logger = logging.getLogger(__name__)
+
+
+@PublicAPI(stability="alpha")
+class TrainingFailedError(RuntimeError):
+    """An error indicating that training has failed."""
+
+    pass
+
+
+@PublicAPI(stability="alpha")
+class Trainer(ConvertibleToTrainable, abc.ABC):
+    """Defines interface for distributed training on Ray.
+
+    Args:
+        scaling_config: Configuration for how to scale training.
+        run_config: Configuration for the execution of the training run.
+        train_dataset: Either a distributed Ray :ref:`Dataset <dataset-api>`
+            or a Callable that returns a Dataset, to use for training. If a
+            ``preprocessor`` is also provided, it will be fit on this
+            dataset and this dataset will be transformed.
+        extra_datasets: Any extra Datasets (such as validation or test
+            datasets) to use for training. If a ``preprocessor`` is
+            provided, the datasets specified here will only be transformed,
```

Comment:
I guess I'll have to see concrete implementations of this class, but I feel like it's a little risky to promise this kind of logic in a pure interface. The interface itself does not reflect any of this (we don't know what "transform" even means here), so subclasses need to take care of it. Maybe this doc string is more suitable for a concrete subclass taking care of that logic.

Reply:
Yep good point, will leave this for the subclasses, but won't mention in the main interface.

Reply:
Ok addressed, ptal!